### ISSUE TITLE
Knowledge transfer for black box optimization

### ISSUE THREAD

**Issue by iaroslav-ai** [scikit-optimize/scikit-optimize#325]
Consider that I want to come up with an algorithm that optimizes treatment parameters  for patients (eg amount & type of medication to take).

Currently if I use algorithms from skopt, I would need to start optimization for particular patient from scratch (correct me if I am wrong). However, I will likely have a lot of data from previous patients, and one can expect it to be useful to converge faster to better solutions. Additionally, for different patients I could provide some features like age etc., which could help even further.

Do you have any plans to include a possibility of knowledge transfer between optimization problems? Maybe something like this: http://papers.nips.cc/paper/5086-multi-task-bayesian-optimization.pdf  .

**Comment by iaroslav-ai** [comment]
One reason why I am interested is that I run many experiments with datasets which are “similar”, to optimize neural network parameters on these datasets. It would be great to reuse data from previous experiments to boost convergence to good neural network architectures on new ones.

**Comment by iaroslav-ai** [comment]
I have been experimenting with this recently. I came up with a code [here](https://gist.github.com/iaroslav-ai/ded34b4e492f1b5dc8e9ddbe3c5cb166), where I have a simplistic   implementation of a modified sequential optimization with GP that can take advantage of similarity of a set of optimization tasks.

Every such “similar” task corresponds to tuning of 3 parameters of SVM on a different synthetic dataset. 

Here is what I get if I do not provide any features describing a task (but data for prior tasks is available):

![figure_1](https://cloud.githubusercontent.com/assets/14038787/23600614/b0ddbaa2-0247-11e7-8f00-ac20d6b33813.png)

Here is what happens if I provide features that tell something abouta a task being solved:

![figure_1-1](https://cloud.githubusercontent.com/assets/14038787/23600623/bd2b03a0-0247-11e7-952d-c547d8a518a3.png)

In both cases I get less of a cold start problem, which is exactly what I am looking for. The data shown in graph is an average for 8 different tasks, repeated 5 times. Knowledge transfer was done for 8 consecutive tasks. 

Here is what I do: I consider a sequence of “similar” tasks as one large optimization task. For every point p and value y that I collect, I transform it into a larger vector by concatenating: 

[f][h][p] = p’

Here f is a vector of task features if any, and h is a one hot encoding of a task number (eg task 3 = [0 0 1 0]). I fit GP regressor to all pairs of p’, y. In order to select a new point to evaluate, I keep f, h constant for particular task, and optimize contents of p in vector p’ wrt acquisition function.

I add vector h to p’ in case every task is independent, so that regressor can distinguish between different optimization tasks. But as it seems regressor can nonetheless take an advantage of a similar structure of different tasks. 

This procedure could be applied with different regressors, not only GP. Similar approach was taken here I think: http://www.cs.ubc.ca/~hutter/papers/10-TR-SMAC.pdf. 

I could come up with PR for skopt implementing this if you would be interested. I am still however not sure about overall procedure. Plus, maybe one can do similar thing by just taking optimal points from other tasks and initializing optimization on new one with these points. 

Would love to know your input on this.

**Comment by betatim** [comment]
Looks very interesting and I'd be interested in a PR. (Caveat emptor: I still owe you a review of your open PR ...). The ideas from the original paper you linked seem sensible etc, but I've never used any of it in the real world.

What would the smallest possible PR be to start towards supporting this? A tool for assembling/dis-assembling the combined feature vector?

**Comment by iaroslav-ai** [comment]
@betatim Yeah, maybe this is something we could start with. I actually have a feeling that this could be integrated into skopt relatively easy, I will look more into skopt to estimate minimal code that could accomplish this. I also want to experiment a bit with current approach on different tasks to see more win / failure cases to understand better how it behaves.

**Comment by iaroslav-ai** [comment]
For those who might be looking into this, I tried the approach outlined above. It is not perfect, and a bit more simple approach seems to perform better (on my SVM tuning optimization problems example). 

While the outlined approach allows to reduce the effect of a cold start problem, the algorithm does not converge to values as good as the algorithm that starts from scratch for every “similar” problem:

![figure_1](https://cloud.githubusercontent.com/assets/14038787/23971916/24ac2fc4-09d0-11e7-847a-6133252b4d21.png)

In the subplot title, "Task number" means that modified knowledge transfer optimizer was only provided with the index of similar task being solved. "Task features and number" means that optimizer was provided with both index of similar task and some features describing the task. The values are the average of best values over 32 repeats for every iteration for 8 similar optimization tasks. 

One could use the knowledge transfer approach for the first few function evaluations and then continue with standard approach in hope to both avoid cold start and not jeopardize the convergence. 

However, one could go in a more simple way about this by trying best points found in previous tasks first, and then proceeding as if the algorithm started from scratch. These are the results that I get (problem and solver code [here](https://gist.github.com/iaroslav-ai/52a83637b4f98287761d5ec043e704ab)):

![reuse_minimum](https://cloud.githubusercontent.com/assets/14038787/23971929/32fb9998-09d0-11e7-9543-a008ff3af2df.png)

The graph above is an average over 100 repeats of the minimum objective values at particular iteration for 10 similar tasks of optimizing 3 parameters of SVM.

Here there is much less of a cold start problem, and the algorithm converges to the same values as the one that starts from scratch. What is also good is that to achieve this one does not require changes in the source code of scikit-optimize.

**Comment by MechCoder** [comment]
Sorry, I cannot comment without having a closer look at the paper linked. I'll try getting there next week.

**Comment by iaroslav-ai** [comment]
I would like to clarify just in case: the non - convergent results are obtained by stacking extra dimensions to the parameter vector as described above. I did not attempt to reproduce results in the paper "Multi-Task Bayesian Optimization".
